### ISSUE TITLE
Migrate build process from CircleCI to GitHub actions

### ISSUE THREAD

**Issue by thomaseizinger** [comit-network/comit-rs#1407]
## Problem

Not having:
- GitHub Action machines are faster
- We can get windows and mac os for free

## Goal

Migrate to GitHub Actions.

## Recommendation

Experiment and consider keeping CircleCI running in parallel.

Benefits:
- Windows, MacOS and Linux builds
- free
- more flexibility

Inspire yourself from the release GitHub Action of comit-rs and the CI GitHub action of create-comit-app.
Note that whatever happens, the release pipeline must stay in CircleCi.

## Out of scope

Making windows build work is out of scope. However, attempting to compile it as part of the CI should.
Do open a follow-up issue for windows.

---

Obsoletes #1051. 
Obsoletes #1098.
Obsoletes #1133. 
Obsoletes #1094.

**Comment by thomaseizinger** [comment]
Can't do this until GitHub actions properly supports caching.

**Comment by thomaseizinger** [comment]
It actually does already!

https://github.com/actions/cache

**Comment by thomaseizinger** [comment]
Blocked by: https://github.com/comit-network/comit-rs/pull/1723

**Comment by thomaseizinger** [comment]
Whoops, wrong button.

**Comment by thomaseizinger** [comment]
The latest release of GitHub actions allows for caches to be up to 2GB!

https://github.com/actions/cache/releases/tag/v1.1.0

**Comment by bonomat** [comment]
@thomaseizinger : what do you think of having a mix between CircleCI and GitHub Actions? For example: 

* Run unix build on CircleCI
* Run macOS build on GitHub Action
* `<kidding>` Ignore windows builds for the rest of our lives  😈`</kidding>` 

Reason is that I don't want to get vendo-locked into GitHub and keep some flexibility.

**Comment by thomaseizinger** [comment]
> @thomaseizinger : what do you think of having a mix between CircleCI and GitHub Actions? For example:
> 
>     * Run unix build on CircleCI
> 
>     * Run macOS build on GitHub Action
> 
>     * `<kidding>` Ignore windows builds for the rest of our lives  smiling_imp`</kidding>`
> 
> 
> Reason is that I don't want to get vendor-locked into GitHub and keep some flexibility.
> 
> //Edit// I propose changing this ticket to create a build for MacOS/Windows using GitHub actions.

The main thing I care about is speed and dev-UX. Some things that are tedious about CircleCI are:

- verbose build scripts
- page load-time

We can go dual build, although I'd hope that GitHub actions allows us to define the build in a way where we literally only have to say: `runs-on: [ubuntu, macos]`.

I understand the vendor-lockin problematic. However, moving _back_ to CircleCI is not that much work that I'd consider it to be a lock-in. 1-2 days and we can migrate back (in case we want)?

**Comment by bonomat** [comment]
Note: when we do that we should introduce a _smart_ way of caching. 
E.g. 
1. build cnd
2. cache cnd
3. yarn install 
4. cache yarn
5. run e2e

**Comment by bonomat** [comment]
Oh no, GiHub action's MacOS does not support Docker :( https://github.community/t5/GitHub-Actions/Why-is-Docker-not-installed-on-macOS/td-p/39364

I assume this was the same reason why travis does not support Docker in their MacOS build environment. 

Since we need Docker for some of our rust tests, I'll exclude MacOS builds from this ticket.

**Comment by thomaseizinger** [comment]
FTR: I am monitoring the issue that adds docker support to the virtual environment: https://github.com/actions/virtual-environments/issues/17

**Comment by thomaseizinger** [comment]
@bonomat FTR: https://github.com/actions/virtual-environments/issues/17 just got closed on the weekend. I did not see a follow-up ticket for enabling mac-os but it should be possible now :)

**Comment by bonomat** [comment]
wohoo... awesome, let's give it a try then :)

**Comment by bonomat** [comment]
@thomas: let's see :) 
https://github.com/comit-network/comit-rs/pull/2345
